Scalar.Packaging.Linux/build.sh

```diff
@@ -0,0 +1,165 @@
+#####################################################################
+echo "Building Scalar.Packaging.Linux..."
+
+# Parse script arguments
+VERSION=$1
+if [ -z "${VERSION}" ]; then
+  echo "Error: Version not specified"
+  exit 1
+fi
+
+CONFIGURATION=$2
+CONFIGURATION="${CONFIGURATION:=Debug}"
+
+# Directories
+THISDIR="$( cd "$(dirname "$0")" ; pwd -P )"
+ROOT="$( cd "$THISDIR"/../.. ; pwd -P )"
+SRC="$( cd "$THISDIR"/.. ; pwd -P )"
+OUT="$ROOT/out"
+SCALAR_SRC="$SRC/Scalar"
+PROJ_OUT="$OUT/Scalar.Packaging.Linux"
+
+# Build parameters
+FRAMEWORK=netcoreapp3.1
+RUNTIME=linux-x64
+
+ARCH="`dpkg-architecture -q DEB_HOST_ARCH`"
+if test -z "$ARCH"; then
+  die "Could not determine host architecture!"
+fi
+
+# Outputs
+PAYLOAD="$PROJ_OUT/payload/$CONFIGURATION"
+SYMBOLOUT="$PROJ_OUT/payload.sym/$CONFIGURATION"
+
+TAROUT="$PROJ_OUT/tar/$CONFIGURATION"
+TARBALL="$TAROUT/scalar-linux_$ARCH.$VERSION.tar.gz"
+SYMTARBALL="$TAROUT/symbols-linux_$ARCH.$VERSION.tar.gz"
+
+DEBOUT="$PROJ_OUT/deb/$CONFIGURATION"
+DEBROOT="$DEBOUT/root"
+DEBPKG="$DEBOUT/scalar-linux_$ARCH.$VERSION.deb"
+DEBROOTAZREPOS="$DEBOUT/root-azrepos"
+DEBPKGAZREPOS="$DEBOUT/scalar-azrepos-linux_$ARCH.$VERSION.deb"
+
+# Cleanup payload directory
+if [ -d "$PAYLOAD" ]; then
+    echo "Cleaning existing payload directory '$PAYLOAD'..."
+    rm -rf "$PAYLOAD"
+fi
+
+# Cleanup symbol directory
+if [ -d "$SYMBOLOUT" ]; then
+    echo "Cleaning existing symbols directory '$SYMBOLOUT'..."
+    rm -rf "$SYMBOLOUT"
+fi
+
+# Ensure directories exists
+mkdir -p "$PAYLOAD" "$SYMBOLOUT" "$DEBROOT" "$DEBROOTAZREPOS"
+
+# Publish core application executables
+echo "Publishing core application..."
+dotnet publish "$SCALAR_SRC" \
+    --configuration="$CONFIGURATION" \
+    --framework="$FRAMEWORK" \
+    --runtime="$RUNTIME" \
+    --self-contained=true \
+    "/p:PublishSingleFile=True" \
+    --output="$(make_absolute "$PAYLOAD")" || exit 1
+
+# Collect symbols
+echo "Collecting managed symbols..."
+mv "$PAYLOAD"/*.pdb "$SYMBOLOUT" || exit 1
+
+echo "Build complete."
+
+#####################################################################
+# PACKING
+#####################################################################
+echo "Packing Scalar.Packaging.Linux..."
+# Cleanup any old archive files
+if [ -e "$TAROUT" ]; then
+    echo "Deleteing old archive '$TAROUT'..."
+    rm "$TAROUT"
+fi
+
+# Ensure the parent directory for the archive exists
+mkdir -p "$TAROUT" || exit 1
+
+# Set full read, write, execute permissions for owner and just read and execute permissions for group and other
+echo "Setting file permissions..."
+/bin/chmod -R 755 "$PAYLOAD" || exit 1
+
+# Build binaries tarball
+echo "Building binaries tarball..."
+pushd "$PAYLOAD"
+tar -czvf "$TARBALL" * || exit 1
+popd
+
+# Build symbols tarball
+echo "Building symbols tarball..."
+pushd "$SYMBOLOUT"
+tar -czvf "$SYMTARBALL" * || exit 1
+popd
+
+# Build .deb
+INSTALL_TO="$DEBROOT/usr/bin/"
+INSTALL_TOAZREPOS="$DEBROOTAZREPOS/usr/bin/"
```

Comment:
But checking the [Filesystem Hierarchy Standard](https://www.debian.org/doc/packaging-manuals/fhs/fhs-3.0.html#theUsrHierarchy) indicates we should be using `/usr/bin` (and that also matches what I see in some of the packages available on https://packages.microsoft.com) so this seems good as-is, but it would be nice to fixup the references in `LinuxPlatform.cs` at the same time.

Reply:
We should fix our `.deb` for `microsoft/git`. Thanks!

Reply:
> But checking the Filesystem Hierarchy Standard indicates we should be using `/usr/bin` [..]

I'm reading that the polar opposite way!

`/usr/local/bin`'s purpose is:

> [..] for use by the system administrator when installing software locally. It needs to be safe from being overwritten when the system software is updated. It may be used for programs and data that are shareable amongst a group of hosts, but not found in /usr.
>
> Locally installed software must be placed within /usr/local rather than /usr unless it is being installed to replace or upgrade software in /usr.

..whereas `/usr/bin`'s purpose is simply:

> [..] is the primary directory of executable commands on the system.

To me this seems to indicate that packages that are being installed by the administrator locally, should be placed in `/usr/local/bin`, but `/usr/bin` is for distribution managed/bundled executables, of which `scalar` is not.

> We should fix our .deb for microsoft/git. Thanks!

I think this is correct. Because `git` _is_ included in most distributions under `/usr/bin`, we should be placing the `git-vfs` (`microsoft/git`) package in `/usr/bin`, because:

> Locally installed software must be placed within /usr/local rather than /usr unless it is being installed to **_replace_** or upgrade software in /usr.

We are looking to replace the bundled Git here, right?

Reply:
If we are installing under `/usr/local/bin`, then that is being picked _before_ `/usr/bin`, so we will be safer installing into `/usr/local/bin`.

The risks I see are:

1. If we install into `/usr/bin`, then if a user installed a different Git version into `/usr/local/bin`, then we will not be executing our version of Git.

2. If we start installing into `/usr/local/bin` but later decide to change into `/usr/bin`, then we will create the above problem for ourselves (and need to un-install from `/usr/local/bin`.

My interpretation of the roles of `/usr/[local/]bin` match @mjcheetham's. However, we are _not_ experts in this area.

Reply:
I agree there is ambiguity in the specification; in particular, it doesn't clearly address third-party packages.

There's a bit more explanation in a Wikipedia [note](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard#cite_note-10):
> Historically and strictly according to the standard, `/usr/local` is for data that must be stored on the local host (as opposed to `/usr`, which may be mounted across a network). Most of the time `/usr/local` is used for installing software/data that are not part of the standard operating system distribution (in such case, `/usr` would only contain software/data that are part of the standard operating system distribution). It is possible that the FHS standard may in the future be changed to reflect this de facto convention.

But are signed third-party packages from repository for which the system admin has agreed to accept and trust downloads (by [adding](https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#2004-) the Microsoft signing key, e.g., in [`InstallFromSource.sh`](https://github.com/microsoft/scalar/blob/09487eeadf282653421fc5fd1eebd933161f8767/Scripts/Linux/InstallFromSource.sh#L32-L33)) part of the "standard distribution"?  A plain language interpretation would suggest not, and argue for `/usr/local/bin`, which was my own first instinct (hence the existing hard-coded path in `LinuxPlatform.cs`), and that's how I manage my own personal local installs of things like .NET Core, FWIW.

But that said, there looks to be a body of opinion that argues `/usr/local` is for "programs not managed by the distribution package manager", as this StackExchange [answer](https://unix.stackexchange.com/a/8658) puts it:
> `/usr/local/bin` is for normal user programs not managed by the distribution package manager, e.g. locally compiled packages. You should not install them into `/usr/bin` because future distribution upgrades may modify or delete them without warning.
And that's more or less how I read the FHS when I looked at it again the other day -- if it's managed by the package manager, it can go into `/usr/bin`, while if you're building it locally, you can override with `/usr/local/bin`.  Admittedly, this is not formalized anywhere I can find.

There's an excellent [comment](https://unix.stackexchange.com/questions/8656/usr-bin-vs-usr-local-bin-on-linux/8658#comment80565_8658) there which points to some of the messy [history](http://lists.busybox.net/pipermail/busybox/2010-December/074114.html).  And it may be worth noting that Debian's own build rules [prohibit](https://lintian.debian.org/tags/file-in-usr-local.html) the use of `/usr/local` in packages.

One could argue `/opt` is really the right place, except that it's not in the default set of paths in the `PATH` environment variable on most systems, so I think that rules it out for us.

The official .NET Core packages appear to use `/usr/share` and `/usr/bin/dotnet` (the latter being a [symlink](https://github.com/dotnet/runtime/blob/34bf5f8d335d1ce8a2966da4b947f753cf95fdf7/src/installer/pkg/packaging/deb/dotnet-sharedhost-debian_config.json#L43) to the actual binary).

All of which just leaves the question open, I guess.  

I feel like the example of the .NET Core packages and at least one way of reading the FHS spec, along with some existing practices, would indicate that we should not be using `/usr/local`.  If someone wanted to build Scalar (or our flavour of Git) from our development source code, they might reasonably expect to be able to put those compiled binaries into `/usr/local/bin` and have them override whatever the system (i.e., the package manager) had installed.

And while it's convenient to say "hey, we want to install our custom version of Git, so let's put it in `/usr/local/bin` since that neatly overrides the Git from the OS vendor", again, if someone was compiling their own Git, they'd expect to be able to use `/usr/local/bin` for it.  And it seems a bit wrong for the package manager to be "stacking" Git installs, putting one "in front of" the other and then counting on the `PATH` environment variable to do the heavy lifting.  It might be more appropriate to use a packaging option such as a [diversion](https://www.debian.org/doc/debian-policy/ap-pkg-diversions.html) or an [alternate version](https://www.debian.org/doc/debian-policy/ap-pkg-alternatives.html) to get the VFS flavour of Git "in front of" the normal one from the OS vendor.

Reply:
For now, I'll go with `/usr/local/bin`. We can revisit this in the future.